src/transaction_builder.h

```diff
@@ -49,6 +50,36 @@ struct OutputDescriptionInfo {
     std::optional<OutputDescription> Build(void* ctx);
 };
 
+struct JSDescriptionInfo {
+    Ed25519VerificationKey joinSplitPubKey;
+    uint256 anchor;
+    std::array<libzcash::JSInput, ZC_NUM_JS_INPUTS> inputs;
+    std::array<libzcash::JSOutput, ZC_NUM_JS_OUTPUTS> outputs;
+    CAmount vpub_old;
+    CAmount vpub_new;
+
+    JSDescriptionInfo(
+        Ed25519VerificationKey joinSplitPubKey,
+        uint256 anchor,
+        std::array<libzcash::JSInput, ZC_NUM_JS_INPUTS> inputs,
+        std::array<libzcash::JSOutput, ZC_NUM_JS_OUTPUTS> outputs,
```

Comment:
Aha, this is the problem. In the previous `JSDescription` constructors, we pased these in by reference; here we clone them. This means that `JSDescriptionInfo::BuildRandomized` is randomizing its copy of the inputs and outputs, not the caller's. The called _does_ still pass `inputMap` and `outputMap` by reference, so it ends up seeing the correct randomised mapping, but when it then creates and stores the Sprout payment disclosure, it looks up the output from its unrandomised `output`.